### ISSUE TITLE
How about to put a hook for version to lifecycle?

### ISSUE THREAD

**Issue by watilde** [npm/npm#7906]
Hi.

I tried to search similar issue but it was not found. So I  decided to propose this.
`npm version` is really convenient and powerful. In order to update the version correctly, I'd like to do that before through to pass the all test and build code. Actually it has already can do that by using `npm-run-script` currently:
```
$ cat package.json
...
scripts: {
  preversion: "npm test & npm run build",
  version: "npm version"
}
...

$ npm run version
```

If I could do that with lifecycle, I think it become more user friendly. Thanks!

**Comment by smikes** [comment]
I like the idea, but one issue is that the `pre` scripts are run unconditionally; they don't really function as guards.  I think a better approach is to write a single lifecycle script , and use that instead of `npm version`:

```
"scripts": {
   "my-version": "npm test && npm run build && npm version"
}
```

Then `npm run my-version` will do what you desire.

Please let me know if I'm misunderstanding your feature request, though!

**Comment by martinheidegger** [comment]
@smikes That script of your example will need to pass all arguments for `my-version` to `npm version` shouldn't it look something like;

```
"scripts": {
   "my-version": "npm test && npm run build && npm version \"$@\""
}
```

My use-case would be to generate screenshots and other things that should be in the versioning system (shown on github) but not generated after every push.

_Note: I was not aware that the scripts don't function as guards, awkward._

**Comment by smikes** [comment]
Hmm, well I don't have a module to test it with, but you should be able to pass parameters after "--" so if "npm version" is the last thing on the line you can use

npm run-script my-version -- args

**Comment by martinheidegger** [comment]
_(Just referenced a duplicate)_

I still think it would be good to have hooks for version. Just for consistency and pragmatic sake. I wonder if they could be made "guarding" in case of `version` only but I am not fixed on it.

**Comment by watilde** [comment]
@smikes Thanks for your comment! Yeah I've known that it can hook using `npm run-script`, but if `npm version` is in the lifecycle it just become more user friendly. And one more thing, I actually think this feature likely helps many people who would like to bump version after passed the all test.

**Comment by bebraw** [comment]
Any update on this?

@watilde Did you figure out something that works? I would love to have proper hooks for `npm version` to keep things simple but I can manage with something simpler.

**Comment by watilde** [comment]
@bebraw Yeah I think so too. I just tried to create a PR last night, will try waiting for a response.

**Comment by zanona** [comment]
That was a great addition —Thanks guys.

I am just wondering that it would be great to be able to know the version number being passed through the script `preversion` for example?

This would be helpful for those having hybrid modules (npm + bower) and bumping versions at the same time?

This could be a nice workflow:

```json
{
    "name": "module",
    "version": "1.0.0",
    "scripts": {
        "preversion": "bower version $1"
    }
}
```

**Comment by jasonkarns** [comment]
I second @zanona's comment. It seems that the `$npm_package_version` variable isn't updated once the actual package.json is bumped. So referencing `$npm_package_version` in the `version` or `postversion` scripts would still give you the old version string. How do we get access to the bumped version without parsing the entire package.json again?

**Comment by smikes** [comment]
This is a great feature request -- would one of you (@jasonkarns or @zanona) mind creating it as a new issue, so it doesn't get lost in the general mass of bugs solved by the new 3.0 release?
